### PR TITLE
use an optimized way to generate a less random content

### PR DESCRIPTION
**Pull request by aryszka**
Signed-off-by: Arpad Ryszka <arpad.ryszka@gmail.com>

### CODE REVIEW

**Comment by szuecs** [comment]
no tests?
Interesting test would be reading more than `len(randomChars)`.

**Comment by szuecs** [comment]
👍

File: filters/diag/diag.go [hunk 1]
```diff
@@ -120,31 +118,15 @@ func (r *random) CreateFilter(args []interface{}) (filters.Filter, error) {
...
-		}
-
-		ni, err := w.Write(b)
-		if err != nil {
-			log.Error("error while writing random content", err)
-			return
-		}
-
-		n += ni
+	if _, err := io.Copy(w, io.LimitReader(r.reader, int64(r.len))); err != nil {
```
Comment (AlexanderYastrebov) [inline filters/diag/diag.go:128]: Is it better than
```go
func (r *random) Request(ctx filters.FilterContext) {
	ctx.Serve(&http.Response{
		StatusCode: http.StatusOK,
		Body:       ioutil.NopCloser(io.LimitReader(r.reader, r.len)),
	})
}
```

File: filters/diag/weakrand.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+var randomChars = []byte("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789")
+
+func weakRandom() *reader {
+	c := make([]byte, len(randomChars))
+	copy(c, randomChars)
+	rand.Shuffle(len(c), func(i, j int) {
+		c[i], c[j] = c[j], c[i]
+	})
+
+	return &reader{c: c, mx: &sync.Mutex{}}
```
Comment (AlexanderYastrebov) [inline filters/diag/weakrand.go:23]: I think there is no need to use mutex pointer

File: filters/diag/weakrand.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+		l := copy(p, r.c[i:])
+		p = p[l:]
+		n += l
+		i += l
+		if i >= len(r.c) {
+			i = 0
+		}
+	}
+
+	r.mx.Lock()
```
Comment (AlexanderYastrebov) [inline filters/diag/weakrand.go:42]: I think we know that returned length is always `len(p)` and also do not need second mutex lock:
```diff
diff --git a/filters/diag/weakrand.go b/filters/diag/weakrand.go
index 4ad7185..4c92ff5 100644
--- a/filters/diag/weakrand.go
+++ b/filters/diag/weakrand.go
@@ -24,23 +24,20 @@ func weakRandom() *reader {
 }
 
 func (r *reader) Read(p []byte) (int, error) {
+       n := len(p)
+
        r.mx.Lock()
        i := r.i
+       r.i = (r.i + n) % len(r.c)
        r.mx.Unlock()
 
-       var n int
        for len(p) > 0 {
                l := copy(p, r.c[i:])
                p = p[l:]
-               n += l
                i += l
-               if i >= len(r.c) {
+               if i == len(r.c) {
                        i = 0
                }
        }
-
-       r.mx.Lock()
-       r.i = i
-       r.mx.Unlock()
        return n, nil
 }
```

**Comment by AlexanderYastrebov** [comment]
Surely gives better performance than https://github.com/zalando/skipper/pull/1570#issuecomment-716853285
```
$ echo "GET http://localhost:9090/" | vegeta attack -duration=1m -rate=0 --max-workers=500 | vegeta report
Requests      [total, rate, throughput]         1058794, 17641.55, 17639.92
Duration      [total, attack, wait]             1m0s, 1m0s, 5.533ms
Latencies     [min, mean, 50, 90, 95, 99, max]  118.308µs, 14.795ms, 11.721ms, 32.977ms, 39.923ms, 55.059ms, 225.154ms
Bytes In      [total, mean]                     8673640448, 8192.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:1058794  
Error Set:
```

File: filters/diag/weakrand.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+type reader struct {
+	c  []byte
+	i  int
+	mx *sync.Mutex
+}
+
+var randomChars = []byte("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789")
+
+func weakRandom() *reader {
```
Comment (AlexanderYastrebov) [inline filters/diag/weakrand.go:16]: It is really weak, better not ask for more than 62 characters :)

Comment (aryszka) [reply filters/diag/weakrand.go:16]: :smile: 

yeah.. So the thing is that we don't want to turn Skipper into a real random generator, we just want to generate some data that is not too predicable for diagnostic purposes. https://xkcd.com/221/

File: filters/diag/weakrand.go (hunk 3 above)
Comment (aryszka) [reply filters/diag/weakrand.go:42]: nice! Thanks a lot!

File: filters/diag/diag.go (hunk 1 above)
Comment (aryszka) [reply filters/diag/diag.go:128]: No, your suggestion is better because it involves less moving parts. Thanks!

**Comment by aryszka** [comment]
@AlexanderYastrebov 

I need to apologize here. I have a couple of hasty days behind me, and I completely missed that you opened a pull request doing the same thing. Sorry!

Let's discuss tomorrow day-time, which approach to go with. Your PR looks good to me, too.

**Comment by aryszka** [comment]
Closing this in favor of: https://github.com/zalando/skipper/pull/1570

**Pull request closed without merging**